### ISSUE TITLE
[Bug] dbt deps sometimes takes >90sec to complete

### ISSUE THREAD

**Issue by joellabes** [dbt-labs/dbt-core#4342]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Over the last couple of weeks, I've noticed that `dbt deps` sometimes takes >90 seconds between `Running with dbt==0.21.0` and the first `Installing...` line.

It's sporadic, and when I re-run to try and capture it performance improves, which makes me suspect it's some sort of cold cache thing. 

### Expected Behavior

<img alt="Let's go faster" src="https://media.giphy.com/media/3ohuPuRnWLwIsCuHPa/giphy.gif">

### Steps To Reproduce

With this packages.yml file (from the internal-analyics project): 
```
packages:
  - local: '../dbt-utils'
  - package: dbt-labs/snowplow
    version: 0.13.0
  - package: dbt-labs/codegen
    version: 0.4.0
  - package: tailsdotcom/dbt_artifacts
    version: 0.5.0
  - package: fivetran/hubspot
    version: 0.4.0
  - package: fivetran/zendesk_source
    version: 0.4.0
  - package: dbt-labs/audit_helper
    version: 0.4.0
```

run `dbt deps`. I was on the VPN, but from memory I've experienced it off the VPN as well. 

Note that this combination will ultimately result in a duplicate package error due to codegen depending on utils as well, but I don't think that's relevant. 

### Relevant log output

```shell
https://www.loom.com/share/95a1c1f62c304673a4649d609aa98781
```

### Environment

```markdown
- OS: macOS 11.6.1
- Python: Python 3.8.10
- dbt: I've noticed this on 0.21.0 as well as various prerelease versions of 1.0
```

### What database are you using dbt with?

snowflake

**Comment by jtcohen6** [comment]
@joellabes Thanks for opening! I've noticed this too.

I believe the significant delay comes from the fact that dbt needs to make _many_ hub API requests in order to "discover" available package versions:
```
2021-11-29 10:25:41.507359 (MainThread): Set downloads directory='/var/folders/7h/hj5_fw9j291c58hwfdvy5xbm0000gp/T/dbt-downloads-nlq7cw6p'
2021-11-29 10:25:41.507605 (MainThread): Making package registry request: GET https://hub.getdbt.com/api/v1/index.json
2021-11-29 10:25:41.786837 (MainThread): Response from registry: GET https://hub.getdbt.com/api/v1/index.json 200
2021-11-29 10:25:41.787967 (MainThread): Making package registry request: GET https://hub.getdbt.com/api/v1/dbt-labs/dbt_utils.json
2021-11-29 10:25:42.027193 (MainThread): Response from registry: GET https://hub.getdbt.com/api/v1/dbt-labs/dbt_utils.json 200
2021-11-29 10:25:42.061386 (MainThread): Making package registry request: GET https://hub.getdbt.com/api/v1/dbt-labs/dbt_utils/0.7.1.json
2021-11-29 10:25:42.266005 (MainThread): Response from registry: GET https://hub.getdbt.com/api/v1/dbt-labs/dbt_utils/0.7.1.json 200
2021-11-29 10:25:42.266890 (MainThread): Making package registry request: GET https://hub.getdbt.com/api/v1/dbt-labs/snowplow.json
2021-11-29 10:25:42.856051 (MainThread): Response from registry: GET https://hub.getdbt.com/api/v1/dbt-labs/snowplow.json 200
... and so on ...
```

(Could you check your debug logs (`logs/dbt.log`), or try running with `dbt --debug deps`, to confirm that's where the delay seems to be?)

My sense is that:
- The hub API can be quite fast or quite slow (perhaps varied based on overall traffic?)
- I don't think this changed at all in v0.21. Anecdotally, I see the same number of hub API requests when I run with v0.20.2 and v0.21.0
- Requests for common dependencies (e.g. `dbt_utils`) are repeated several times. We could probably find ways to reduce the number of requests that dbt-core is making to the Hub API, or—though this is a heavier lift—find ways to make those requests in parallel

**Comment by joellabes** [comment]
@jtcohen6 yes there seems to be 1min 38sec of registry requests (between 2:18:49 and 2:20:27) before any package is installed - [dbt.log](https://github.com/dbt-labs/dbt-core/files/7885717/dbt.log). 2min 5 sec before the `upgrades available` message is displayed at 2:20:54. 

Agreed that this isn't a 1.0 regression. it had been annoying me for a while but I hadn't gotten around to documenting it. 

Parallel requests would be interesting for the installs towards the end, but – at least for our project with 7 packages – it seems like the real win would be if the hub requests could be a batch request. If we could pass up a request containing all the Hub-hosted packages at once, and get back an array of tarballs + hashes then that'd be nicer. 

I just had a bit of a look at the objects being returned from the [hub API](https://hub.getdbt.com/api/v1/tailsdotcom/dbt_artifacts.json), and it looks like that sort of change would mean all the version and dependency logic would start to live on the server instead of on the client. Do we have an appetite for that, or is it a *feature* that the client does all this locally? (or is this what you meant when you said parallelising would be a heavier lift)

**Comment by jtcohen6** [comment]
I don't feel so strongly about where version/dependency resolution happens. It's a bit strange that the logic for it lives in `dbt-core`; it might be a little bit stranger if it lived within hub.getdbt.com. We've talked about splitting this out into a separate package (`dbt-deps`?), anticipating a future in which we need its capabilities in multiple places and want to avoid duplicating code.

From a functional standpoint, if all version/dependency resolution were to happen via a single batch call to the Hub API, what would this look like? Send up the full contents of `packages.yml`, and receive back the link to all relevant tarballs? I'm worried that this might not work for package specifications that mix `local` packages, `git` packages, and Hub-hosted `package` entries—thinking, too, that the first two can declare their own dependencies on the third.

**Comment by joellabes** [comment]
> Send up the full contents of `packages.yml`, and receive back the link to all relevant tarballs

I was thinking just the ones identified as hub-hosted, but whether it's discarded client-side or server-side doesn't _really_ matter.



> the first two can declare their own dependencies on the third

But i had forgotten about this wrinkle. Is the same true in reverse (a hub package declaring a git dependency)? If not, then maybe we resolve all of those first, then we know the whole list of hub packages we need to dig up.

**Comment by jeremyyeo** [comment]
Adding this here as @joellabes pointed out there's a related issue. Almost a good 5 minutes to install the [ad_reporting](https://hub.getdbt.com/fivetran/ad_reporting/latest/) super package.

https://www.loom.com/share/b8663b40c1d548ff91fe8cf04042a670

**Comment by dbeatty10** [comment]
Fixed by #4982
